File: algorithm/problem/insertingSort-1/answer.js
// answer 1 ------------------------------------
function insertionSort1(n, arr) {
  let finish = false;
  const lastValue = arr[n - 1];
  const subArray = arr.slice(0, n - 1);

  for (let i = subArray.length - 1; i >= -1; i--) {
    // handle index 0 comparison
    if (i < 0 || subArray[i] < lastValue) {
      subArray.splice(i + 1, 1, lastValue);
      finish = true;
    } else {
      subArray.splice(i + 1, 1, subArray[i]);
    }

    console.log(subArray.join(" "));
    if (finish) break;
  }
}

// answer 2 ------------------------------------
function insertionSort2(n, arr) {
  target = arr[n - 1];
  idx = n - 2;

  while (idx >= 0 && target < arr[idx]) {
    arr[idx + 1] = arr[idx];
    console.log(arr.join(" "));
    idx -= 1;
  }

  arr[idx + 1] = target;
  console.log(arr.join(" "));
}

// answer 3 ------------------------------------
// Note: below solution modify given array instead of create anew

const insert = (array, rightIndex, value) => {
  for (var j = rightIndex; j >= 0 && array[j] > value; j--) {
    array[j + 1] = array[j];
  }

  array[j + 1] = value;
};

const insertionSort = (array) => {
  for (var i = 1; i < array.length; i++) {
    insert(array, i - 1, array[i]);
  }
};

var array = [22, 11, 99, 88, 9, 7, 42];
insertionSort(array);

File: algorithm/problem/insertingSort-1/answer.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    static void insertionSort1(int n, int[] arr) {    
      int lastValue = arr[n - 1];
      boolean finish = false;
      
      for (int i = n - 2; i >= -1; i--) {
        if (i < 0 || arr[i] < lastValue) {
          arr[i + 1] = lastValue;
          finish = true;
        } else {
          arr[i + 1] = arr[i];
        }
        
        String[] strArray = Arrays.stream(arr).mapToObj(String::valueOf).toArray(String[]::new);
        String joinedString = String.join(" ", strArray);
        System.out.println(joinedString);
        if (finish) break;
      }
    }
}
